File: backend/routes/p6Token.js
// P6 Token Routes
const express = require('express');
const router = express.Router();
const { generateP6Token, getValidP6Token, clearCachedToken, isTokenValid } = require('../services/p6TokenService');

// Authentication middleware will be passed from server.js
let authenticateToken;

const setPool = (pool, authMiddleware) => {
    authenticateToken = authMiddleware;
};

/**
 * POST /api/p6-token/generate
 * Generate a fresh P6 OAuth token (called on user login)
 */
router.post('/generate', (req, res, next) => authenticateToken(req, res, next), async (req, res) => {
    try {
        const tokenData = await generateP6Token();
        res.json({
            success: true,
            message: 'P6 token generated successfully',
            token: tokenData.accessToken,
            expiresIn: tokenData.expiresIn
        });
    } catch (error) {
        console.error('Error generating P6 token:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to generate P6 token',
            error: error.message
        });
    }
});

/**
 * GET /api/p6-token/current
 * Get the current valid token (uses cache or generates new one)
 */
router.get('/current', (req, res, next) => authenticateToken(req, res, next), async (req, res) => {
    try {
        const token = await getValidP6Token();
        res.json({
            success: true,
            token: token,
            isValid: isTokenValid()
        });
    } catch (error) {
        console.error('Error getting P6 token:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to get P6 token',
            error: error.message
        });
    }
});

/**
 * POST /api/p6-token/refresh
 * Force refresh the P6 token
 */
router.post('/refresh', (req, res, next) => authenticateToken(req, res, next), async (req, res) => {
    try {
        clearCachedToken();
        const tokenData = await generateP6Token();
        res.json({
            success: true,
            message: 'P6 token refreshed successfully',
            token: tokenData.accessToken,
            expiresIn: tokenData.expiresIn
        });
    } catch (error) {
        console.error('Error refreshing P6 token:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to refresh P6 token',
            error: error.message
        });
    }
});

/**
 * GET /api/p6-token/status
 * Check token status
 */
router.get('/status', (req, res, next) => authenticateToken(req, res, next), (req, res) => {
    res.json({
        success: true,
        isValid: isTokenValid()
    });
});

module.exports = { router, setPool };

File: backend-fastapi/app/routers/p6_token.py
# app/routers/p6_token.py
"""
P6 Token router.
Direct port of Express routes/p6Token.js
"""

from fastapi import APIRouter, Depends

from app.auth.dependencies import get_current_user
from app.services.p6_token_service import (
    generate_p6_token,
    get_valid_p6_token,
    is_token_valid,
    clear_cached_token,
)

router = APIRouter(prefix="/api/p6-token", tags=["P6 Token"])


@router.post("/generate")
async def generate_token(current_user: dict = Depends(get_current_user)):
    """Generate a new Oracle P6 token."""
    try:
        token = await generate_p6_token()
        return {"success": True, "message": "Token generated successfully", "token": token}
    except Exception as e:
        return {"success": False, "message": f"Failed to generate token: {e}"}


@router.get("/current")
async def get_current_token(current_user: dict = Depends(get_current_user)):
    """Get the current valid P6 token."""
    try:
        token = await get_valid_p6_token()
        return {"success": True, "token": token}
    except Exception as e:
        return {"success": False, "message": f"No valid token available: {e}"}


@router.post("/refresh")
async def refresh_token(current_user: dict = Depends(get_current_user)):
    """Force refresh the P6 token."""
    clear_cached_token()
    try:
        token = await generate_p6_token()
        return {"success": True, "message": "Token refreshed successfully", "token": token}
    except Exception as e:
        return {"success": False, "message": f"Failed to refresh token: {e}"}


@router.get("/status")
async def get_token_status(current_user: dict = Depends(get_current_user)):
    """Check if the cached P6 token is valid."""
    return {"valid": is_token_valid()}
